cpp/src/row_conversion/row_conversion.cu

```diff
@@ -0,0 +1,1321 @@
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <iostream>
+#include <iterator>
+#include <limits>
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/sequence.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <cudf/row_conversion.hpp>
+#include "cudf/types.hpp"
+#include "rmm/device_buffer.hpp"
+#include "thrust/iterator/counting_iterator.h"
+#include "thrust/iterator/transform_iterator.h"
+
+namespace cudf {
+
+namespace detail {
+
+static inline __host__ __device__ int32_t align_offset(int32_t offset, std::size_t alignment)
+{
+  return (offset + alignment - 1) & ~(alignment - 1);
+}
+
+/**
+ * Copy a simple vector to device memory asynchronously. Be sure to read
+ * the data on the same stream as is used to copy it.
+ */
+template <typename T>
+std::unique_ptr<rmm::device_uvector<T>> copy_to_dev_async(const std::vector<T> &input,
+                                                          rmm::cuda_stream_view stream,
+                                                          rmm::mr::device_memory_resource *mr)
+{
+  std::unique_ptr<rmm::device_uvector<T>> ret(new rmm::device_uvector<T>(input.size(), stream, mr));
+  CUDA_TRY(cudaMemcpyAsync(
+    ret->data(), input.data(), sizeof(T) * input.size(), cudaMemcpyHostToDevice, stream.value()));
+  return ret;
+}
+
+template <typename T>
+rmm::device_uvector<T> copy_to_dev_async2(const std::vector<T> &input,
+                                          rmm::cuda_stream_view stream,
+                                          rmm::mr::device_memory_resource *mr)
+{
+  rmm::device_uvector<T> ret(input.size(), stream, mr);
+  CUDA_TRY(cudaMemcpyAsync(
+    ret.data(), input.data(), sizeof(T) * input.size(), cudaMemcpyHostToDevice, stream.value()));
+  return ret;
+}
+
+__global__ void copy_to_fixed_width_columns(const cudf::size_type num_rows,
+                                            const cudf::size_type num_columns,
+                                            const cudf::size_type row_size,
+                                            const cudf::size_type *input_offset_in_row,
+                                            const cudf::size_type *num_bytes,
+                                            int8_t **output_data,
+                                            cudf::bitmask_type **output_nm,
+                                            const int8_t *input_data)
+{
+  // We are going to copy the data in two passes.
+  // The first pass copies a chunk of data into shared memory.
+  // The second pass copies that chunk from shared memory out to the final location.
+
+  // Because shared memory is limited we copy a subset of the rows at a time.
+  // For simplicity we will refer to this as a row_group
```

Comment:
This would be a perfect algorithm for `memcpy_async` and setting up a 2 stage pipeline of copying in data to shared memory asynchronously while writing out the other stage. See https://developer.nvidia.com/blog/controlling-data-movement-to-boost-performance-on-ampere-architecture/